### PR TITLE
Dynamic Input, hooks, backward compatibility with topologies

### LINKED ISSUE #289: Partial completion of beaker provision results in no inventory file
Provisioning multiple machines, all but one pass with one failure and there is no inventory file created for the ones that passed. For us, this means we either have to build the inventory file by hand or redo the provisioning and hope it passes.

topology file:
```
---
topology_name: beaker_server
resource_groups:
  - resource_group_name: pod_node1
    resource_group_type: beaker
    job_group: "CS&S"
    recipesets:
      - distro: RHEL-7.3-updates-20170608.0
        arch: x86_64
        count: 1
        hostrequires:
          - tag: pool
            op: "="
            value: "CS&S"
          - tag: hostname
            op: like
            value: "dell-per530%"
  - resource_group_name: pod_node2
    resource_group_type: beaker
    job_group: "CS&S"
    recipesets:
      - distro: RHEL-7.3-updates-20170608.0
        arch: x86_64
        count: 1
        hostrequires:
          - tag: pool
            op: "="
            value: "CS&S"
          - tag: hostname
            op: like
            value: "dell-per530%"
  - resource_group_name: pod_node3
    resource_group_type: beaker
    job_group: "CS&S"
    recipesets:
      - distro: RHEL-7.3-updates-20170608.0
        arch: x86_64
        count: 1
        hostrequires:
          - tag: pool
            op: "="
            value: "CS&S"
          - tag: hostname
            op: like
            value: "dell-per530%"
  - resource_group_name: pod_node4
    resource_group_type: beaker
    job_group: "CS&S"
    recipesets:
      - distro: RHEL-7.3-updates-20170608.0
        arch: x86_64
        count: 1
        hostrequires:
          - tag: pool
            op: "="
            value: "CS&S"
          - tag: hostname
            op: like
            value: "dell-per530%"
  - resource_group_name: pod_node5
    resource_group_type: beaker
    job_group: "CS&S"
    recipesets:
      - distro: RHEL-7.3-updates-20170608.0
        arch: x86_64
        count: 1
        hostrequires:
          - tag: pool
            op: "="
            value: "CS&S"
          - tag: hostname
            op: like
            value: "dell-per530%"
  - resource_group_name: pod_node6
    resource_group_type: beaker
    job_group: "CS&S"
    recipesets:
      - distro: RHEL-7.3-updates-20170608.0
        arch: x86_64
        count: 1
        hostrequires:
          - tag: pool
            op: "="
            value: "CS&S"
          - tag: hostname
            op: like
            value: "dell-per530%"
  - resource_group_name: pod_node7
    resource_group_type: beaker
    job_group: "CS&S"
    recipesets:
      - distro: RHEL-7.3-updates-20170608.0
        arch: x86_64
        count: 1
        hostrequires:
          - tag: pool
            op: "="
            value: "CS&S"
          - tag: hostname
            op: like
            value: "dell-per530%"
  - resource_group_name: pod_node8
    resource_group_type: beaker
    job_group: "CS&S"
    recipesets:
      - distro: RHEL-7.3-updates-20170608.0
        arch: x86_64
        count: 1
        hostrequires:
          - tag: pool
            op: "="
            value: "CS&S"
          - tag: hostname
            op: like
            value: "dell-per530%"
```

One of these fails, 7 pass, but no inventory file for those that pass.

### LINKED ISSUE #402: Dynamic inputs to linchpin
Dynamic PinFile, topology, and layout files. Essentially, the idea would be that a user could pass in a script in place of these items. The result of the script would be a json output, probably to STDOUT, that linchpin could ingest.

Probably something like:

```
[{
    "topology_data": {
        "resource_groups": [{
            "credentials": {
                "filename": "clouds.yaml",
                "profile": "ci-rhos"
            },
            "resource_definitions": [{
                "count": 1,
                "fip_pool": "10.8.240.0",
                "flavor": "m1.small",
                "image": "CentOS-7-x86_64-GenericCloud-1612",
                "keypair": "ci-factory",
                "name": "database",
                "networks": [
                    "atomic-e2e-jenkins-test"
                ],
                "type": "os_server"
                },
                {
                "count": 1,
                "fip_pool": "10.8.240.0",
                "flavor": "m1.small",
                "image": "Fedora-Cloud-Base-26-compose-latest",
                "keypair": "ci-factory",
                "name": "frontend",
                "networks": [
                    "atomic-e2e-jenkins-test"
                ],
                "type": "os_server"
                }
            ],
            "resource_group_name": "os-server",
            "resource_group_type": "openstack"
        }],
        "topology_name": "os-single"
    }
},
{
    "layout_data": {
        "inventory_layout": {
            "hosts": {
                "db-node": {
                    "count": 1,
                    "host_groups": [
                        "database"
                    ]
                },
                "frontend-node": {
                    "count": 1,
                    "host_groups": [
                        "frontend"
                    ]
                }
            },
            "vars": {
                "hostname": "__IP__"
            }
        }
    }
}]
```

### PR DESCRIPTION
**Pull request by herlo**
A dynamic `PinFile` (which include topology and/or layout options) can now be used. A script can be passed which would normally run on the command line, which outputs JSON to STDOUT. The JSON must describe a standard PinFile. An example of the output may look like this:

```
{
    "dummy": {
        "topology": {
            "resource_groups": [
                {
                    "resource_definitions": [
                        {
                            "count": 3,
                            "name": "web",
                            "role": "dummy_node"
                        }
                    ],
                    "resource_group_name": "dummy",
                    "resource_group_type": "dummy"
                }
            ],
            "topology_name": "dummy_cluster"
        }
    }
}
```
But also may look like this:

```
{
    "dummy": {
        "topology": "dummy.yml",
        "layout": "dummy.yml"       
    }
}
```

To pass this script, one would use the linchpin CLI like so:

`linchpin -vp /path/to/script up [target]` 

Additionally, the PinFile can represent both the topology and layouts in YAML as before. Plus, now supports inline topology and layouts as so:

```
dummy-dyn:
  topology:
    topology_name: "dummy_cluster" # topology name
    resource_groups:
      - resource_group_name: "dummy"
        resource_group_type: "dummy"
        resource_definitions:
          - name: "web"
            role: "dummy_node"
            count: 3
          - name: "test"
            role: "dummy_node"
            count: 1
  layout:
    inventory_layout:
      vars:
        hostname: __IP__
      hosts:
        example-node:
          count: 3
          host_groups:
            - example
        test-node:
          count: 1
          host_groups:
            - test
  hooks:
    postup:
      - type: shell
        name: hello
        actions:
          - echo hello
```

Considering the new topology schema constraints, old topologies will now work for the most part with the translator in place. 

Hooks work well again, with some minor improvements.

Also, during testing, it was discovered that the `default_inventory_path` was being used in the inventory_gen role. This meant that every inventory generated would be the same, and the inventory_layout was being ignored. To remedy this, the InventoryFilter.py was updated. In addition to addressing the bug, inventories now generate properly, giving way to Fixing #289.

Because the inventory_layout data is now being stored in an extra_var, the ordered_yaml filter was also cleaned up some.

Fixes #402 
Fixes #289

### CODE REVIEW

File: linchpin/__init__.py [hunk 1]
```diff
@@ -229,6 +231,28 @@ def _find_playbook_path(self, playbook):
...
+        try:
+            for res_grp in topology.get('resource_groups'):
+                if 'res_group_type' in res_grp.keys():
+                    res_grp['resource_group_type'] = (
+                        res_grp.pop('res_group_type'))
+                for res_def in res_grp.get('resource_definitions'):
+                    if 'res_name' in res_def.keys():
+                        res_def['name'] = res_def.pop('res_name')
+                    if 'type' in res_def.keys():
+                        res_def['role'] = res_def.pop('type')
```
[the remark this answers is not in the document]
Comment (herlo) [reply linchpin/__init__.py]: @p3ck funny you should mention this. I just noticed it today and update to fix exactly that.

[the remark this answers is not in the document]
Comment (herlo) [reply linchpin/__init__.py]: @p3ck I mean, I have an update to fix exactly that. It will come in the next PR, if that's acceptable.

**Comment by herlo** [comment]
@p3ck I see the comment about the inventory_gen error. Can you clarify here whether that is specific to duffy or not? I read 'All targets fail duffy provisioning with the following error'. What does that mean exactly? Do all targets fail, or do all duffy targets fail?

**Comment by herlo** [comment]
@p3ck I just updated to this branch on centos CI server and ran a duffy target. I am unable to replicate the inventory_gen issue.

File: linchpin/__init__.py [hunk 2]
```diff
@@ -229,6 +231,40 @@ def _find_playbook_path(self, playbook):
...
+                    res_grp.pop('res_defs'))
+
+            for res_def in res_grp.get('resource_definitions'):
+                if res_def:
+                    if 'res_name' in res_def.keys():
+                        res_def['name'] = res_def.pop('res_name')
+                    if 'type' in res_def.keys():
+                        res_def['role'] = res_def.pop('type')
+                    if 'res_type' in res_def.keys():
+                        res_def['role'] = res_def.pop('res_type')
```
Comment (herlo) [inline linchpin/__init__.py]: Decided to address this today. Here's the fix.

**Comment by p3ck** [comment]
@herlo I checked out pr422 and did a make test in Dockerfiles and Duffy failed on all distro targets (f25, f26, f27 and Centos7)

**Comment by herlo** [comment]
@p3ck The big difference I could see was that the include was not checking for layout_data properly. The new version in duffy.yml does this and appears to be working. Please test this again and let me know if it continues to fail.

**Pull request merged**